addons/account/account.py

```diff
@@ -269,15 +269,33 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None,
                 order, context=context, count=count)
 
     def _get_children_and_consol(self, cr, uid, ids, context=None):
-        #this function search for all the children and all consolidated children (recursively) of the given account ids
-        ids2 = self.search(cr, uid, [('parent_id', 'child_of', ids)], context=context)
-        ids3 = []
-        for rec in self.browse(cr, uid, ids2, context=context):
-            for child in rec.child_consol_ids:
-                ids3.append(child.id)
-        if ids3:
-            ids3 = self._get_children_and_consol(cr, uid, ids3, context)
-        return ids2 + ids3
+        if isinstance(ids, (int, long)):
+            ids = [ids]
+        cr.execute(
+            """
+            SELECT id FROM (
+                WITH RECURSIVE account_children(id, depth) AS (
+                        SELECT id, 0 as depth
+                        FROM account_account
+                        WHERE id IN %s
+                UNION
+                        SELECT a.id, ac.depth + 1
+                        FROM account_account a, account_children ac
+                        WHERE a.parent_id = ac.id OR a.id IN
+                                (SELECT parent_id
+                                 FROM account_account_consol_rel a
+                                 WHERE a.child_id = ac.id)
+                )
+                SELECT DISTINCT id, depth
+                     , max(depth) OVER (PARTITION BY id) as max_depth
```

Comment:
I usually format my queries like this, when selecting many fields:
```
SELECT foo
            , bar
            , baz
FROM spam
```
because then I make less errors with trailing commas, especially when reordering fields. Just personal style.